src/NHibernate/Criterion/EntityProjection.cs

```diff
@@ -0,0 +1,175 @@
+		public bool Lazy { get; set; }
+
+		internal string[] IdentifierColumnAliases { get; private set; }
+		internal string[][] PropertyColumnAliases { get; private set; }
+		internal IQueryable Persister { get; private set; }
+		internal System.Type RootEntity { get; private set; }
+
+		#region Configuration methods
+
+		public TThis SetLazy(bool lazy = true)
+		{
+			Lazy = lazy;
+			return this;
+		}
+
+		public TThis SetAllPropertyFetch(bool fetchLazyProperties = true)
+		{
+			FetchLazyProperties = fetchLazyProperties;
+			return this;
+		}
+
+		public TThis SetReadonly(bool isReadOnly = true)
+		{
+			IsReadOnly = isReadOnly;
+			return this;
+		}
+
+		#endregion Configuration methods
+
+		#region IProjection implementation
+
+		/// <summary>
+		/// Entity alias
+		/// </summary>
+		public string[] Aliases
+		{
+			get { return new[] {_entityAlias}; }
+		}
+
+		bool IProjection.IsAggregate
+		{
+			get { return false; }
+		}
+
+		bool IProjection.IsGrouped
+		{
+			get { return false; }
+		}
+
+		IType[] IProjection.GetTypes(string alias, ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			return new[] {criteriaQuery.GetType(criteria, alias)};
+		}
+
+		IType[] IProjection.GetTypes(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			return _types;
+		}
+
+		public string[] GetColumnAliases(int position, ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			return _columnAliases;
+		}
+
+		public string[] GetColumnAliases(string alias, int position, ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			return _columnAliases;
+		}
+
+		SqlString IProjection.ToSqlString(ICriteria criteria, int position, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			string identifierSelectFragment = Persister.IdentifierSelectFragment(_tableAlias, _columnAliasSuffix);
+			return new SqlString(
+				Lazy
+					? identifierSelectFragment
+					: string.Concat(
+						identifierSelectFragment,
+						Persister.PropertySelectFragment(_tableAlias, _columnAliasSuffix, FetchLazyProperties)));
+		}
+
+		SqlString IProjection.ToGroupSqlString(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			throw new NotImplementedException();
+		}
+
+		TypedValue[] IProjection.GetTypedValues(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			throw new NotImplementedException();
+		}
+
+		#endregion IProjection implementation
+
+		private void SetFields(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			//Persister is required, so let's use it as "initialized marker"
+			if (Persister != null)
+				return;
+
+			if (RootEntity == null)
+			{
+				RootEntity = criteria.GetRootEntityTypeIfAvailable();
+			}
+
+			if (_entityAlias == null)
+			{
+				_entityAlias = criteria.Alias;
+			}
+
+			Persister = (IQueryable) criteriaQuery.Factory.GetEntityPersister(RootEntity.FullName);
```

Comment:
This cast will fail if a custom persister is used. The code should handle more gracefully this case, maybe with something like:
```c#
Persister = criteriaQuery.Factory.GetEntityPersister(RootEntity.FullName) as IQueryable;
if (Persister == null)
    throw new InvalidOperationException($"Projecting to entities requires a NHibernate.Persister.Entity.IQueryable persister, {RootEntity.FullName} does not have one.");

Reply:
Ok.

For some reasons I can't comment under your previous comment. So answering here:
>In fact we favor putting tests in their functional area, so it would be even better to move them to Criteria/ProjectionsTest.cs.

I didn't do it because I have my set of entities specifically for my needs and I don't like hbm configuration (and it's not clear to me how to mix them). It doesn't look to me as clean solution to mix all sort of various tests depending on different data in one file and do set up/clean up for all of them for each test method.  I would really prefer keep my tests in separate class
Maybe it's better to introduce Criteria/Projections folder? And put files there (for now two files - ProjectionsTest.cs and new EntityProjectionTest.cs)

Reply:
I have slightly simplified my proposed code in previous comment.

> For some reasons I can't comment under your previous comment.

I have seen this happening on the conversation page, where comments appears somewhat regrouped by their respective review, and may appears in subsequent ones too. You have to find the "right" group of comments. Or switch to files tab and answer from there.